app/geant-exporter/GeantLoggerAdapter.cc

```diff
@@ -0,0 +1,82 @@
+//----------------------------------*-C++-*----------------------------------//
+// Copyright 2020 UT-Battelle, LLC, and other Celeritas developers.
+// See the top-level COPYRIGHT file for details.
+// SPDX-License-Identifier: (Apache-2.0 OR MIT)
+//---------------------------------------------------------------------------//
+//! \file GeantLoggerAdapter.cc
+//---------------------------------------------------------------------------//
+#include "GeantLoggerAdapter.hh"
+
+#include <cctype>
+#include <G4strstreambuf.hh>
+#include "comm/Logger.hh"
+
+namespace celeritas
+{
+//---------------------------------------------------------------------------//
+/*!
+ * Assign exception handler on construction.
+ */
+GeantLoggerAdapter::GeantLoggerAdapter()
+    : saved_cout_(G4coutbuf.GetDestination())
+    , saved_cerr_(G4cerrbuf.GetDestination())
+{
+    G4coutbuf.SetDestination(this);
+    G4cerrbuf.SetDestination(this);
+}
+
+//---------------------------------------------------------------------------//
+/*!
+ * Restore exception handler on destruction.
+ */
+GeantLoggerAdapter::~GeantLoggerAdapter()
+{
+    if (G4coutbuf.GetDestination() == this)
+    {
+        G4coutbuf.SetDestination(saved_cout_);
+    }
+    if (G4cerrbuf.GetDestination() == this)
+    {
+        G4cerrbuf.SetDestination(saved_cerr_);
+    }
+}
+
+//---------------------------------------------------------------------------//
+/*!
+ * Process stdout message.
+ */
+G4int GeantLoggerAdapter::ReceiveG4cout(const G4String& str)
+{
+    return this->log_impl(str, LogLevel::diagnostic);
+}
+
+//---------------------------------------------------------------------------//
+/*!
+ * Process stderr message.
+ */
+G4int GeantLoggerAdapter::ReceiveG4cerr(const G4String& str)
+{
+    return this->log_impl(str, LogLevel::info);
+}
+
+//---------------------------------------------------------------------------//
+/*!
+ * Log the actual message.
+ */
+G4int GeantLoggerAdapter::log_impl(const G4String& str, LogLevel level)
+{
+    // Strip trailing whitespace
+    while (!str.empty() && std::isspace(static_cast<unsigned char>(str.back())))
+    {
+        const_cast<G4String&>(str).pop_back();
```

Comment:
Why is the const_cast preferable to a copy? ... well obviously it is 'faster' but is the performance gain worth the encapsulation break?

Reply:
I based this code off some of the older stuff I'd put in exnihilo, including this "optimization". You're right that it's better not to break encapsulation, though that's a laughable goal when involving geant4 ;) I'll update to use a copy to be cleaner.

Reply:
More than encapsulation that would protect against somewhat unexpected behavior ( either the string being really really const or (more likely) being reused and assumed unchanged).